### ISSUE TITLE
External storage is scanned for each user during reindexing

### ISSUE THREAD

**Issue by w1ndy** [owncloud/core#32210]
### Steps to reproduce
1. Set up external storage shared by multiple users
2. Execute ``occ files:scan -v --all``

### Expected behaviour
The external storage should be scanned only once.

### Actual behaviour
The external storage is being scanned for every user (e.g. /user1/Data/../., /user2/Data/../., /user3/Data/../. are all scanned when Data is external storage), which makes the reindexing incredibly slow.

```
+---------+--------+--------------+
| Folders | Files  | Elapsed time |
+---------+--------+--------------+
| 21721   | 191613 | 00:11:11     |
+---------+--------+--------------+
```

### Server configuration
**Operating system**:
Docker on CentOS 7.3

**Web server:**
NGINX

**Database:**
MariaDB:latest

**PHP version:**
Docker image default

**ownCloud version:** (see ownCloud admin page)
10.0.9.5

**Updated from an older ownCloud or fresh install:**
Fresh install

**Where did you install ownCloud from:**
Docker

**Signing status (ownCloud 9.0 and above):**

```
No errors have been found.
```

**The content of config/config.php:**

```
{
    "system": {
        "datadirectory": "\/mnt\/data\/files",
        "apps_paths": [
            {
                "path": "\/var\/www\/owncloud\/apps",
                "url": "\/apps",
                "writable": false
            },
            {
                "path": "\/var\/www\/owncloud\/custom",
                "url": "\/custom",
                "writable": true
            }
        ],
        "dbtype": "mysql",
        "dbhost": "db:3306",
        "dbname": "owncloud",
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "dbtableprefix": "oc_",
        "trusted_domains": [
            "localhost"
        ],
        "mysql.utf8mb4": true,
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "overwrite.cli.url": "http:\/\/cloud.zjuvis.org\/",
        "version": "10.0.9.5",
        "logtimezone": "UTC",
        "installed": true,
        "instanceid": "oci4cvp0z0oe",
        "updatechecker": "false",
        "upgrade.disable-web": "true",
        "redis": {
            "host": "redis",
            "port": 6379
        },
        "memcache.distributed": "\\OC\\Memcache\\Redis",
        "memcache.locking": "\\OC\\Memcache\\Redis",
        "filelocking.enabled": "true",
        "memcache.local": "\\OC\\Memcache\\APCu",
        "maintenance": false,
        "loglevel": "2",
        "default_language": "en",
        "htaccess.RewriteBase": "\/",
        "mail_domain": "***REMOVED SENSITIVE VALUE***",
        "mail_from_address": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpmode": "smtp",
        "mail_smtphost": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpport": "25",
        "files_external_allow_create_new_local": "true"
    }
}
```

**List of activated apps:**

```
Enabled:
  - activity: 2.3.7
  - calendar: 1.5.7
  - comments: 0.3.0
  - configreport: 0.1.1
  - contacts: 1.5.4
  - dav: 0.3.2
  - federatedfilesharing: 0.3.1
  - federation: 0.1.0
  - files: 1.5.1
  - files_clipboard: 0.7.1
  - files_external: 0.7.1
  - files_pdfviewer: 0.9.0
  - files_sharing: 0.10.1
  - files_texteditor: 2.2.1
  - files_trashbin: 0.9.1
  - files_versions: 1.3.0
  - files_videoplayer: 0.9.8
  - firstrunwizard: 1.1
  - market: 0.2.4
  - notes: 2.0.4
  - notifications: 0.3.4
  - provisioning_api: 0.5.0
  - systemtags: 0.3.0
  - tasks: 0.9.6
  - updatenotification: 0.2.1
Disabled:
  - audioplayer
  - encryption
  - external
  - gallery
  - oauth2
  - onlyoffice
  - user_external
  - user_ldap
```

**Are you using external storage, if yes which one:** local/smb/sftp/...
Local

**Are you using encryption:** yes/no
No

**Are you using an external user-backend, if yes which one:** LDAP/ActiveDirectory/Webdav/...
No

**Comment by ownclouders** [comment]
[GitMate.io](https://gitmate.io) thinks the contributor most likely able to help you is @mmattel.

Possibly related issues are https://github.com/owncloud/core/issues/10198 (External Only Storage), https://github.com/owncloud/core/issues/6682 (External Storage - SMB), https://github.com/owncloud/core/issues/17943 (files:scan for a user with only external storage scans ALL files), https://github.com/owncloud/core/issues/21474 (External storage failed), and https://github.com/owncloud/core/issues/23277 (disable external storages).

**Comment by mmattel** [comment]
hello @w1ndy 

``files:scan --help`` tells you:
``--all              Will rescan all files of all known users.``

The documentation of the occ command also tells you the following:

a.) using the --path option only scans you the given folder/mountpoint of a user defined in path
and
b.) Mounts are only scannable at the point of origin. Scanning of shares including federated shares is not necessary on the receiver side and therefore not possible

From my understanding of what you wrote above:
Set up external storage shared by multiple users
means to me that a user or the admin creates a mount point and shares it to multiple users/groups. which can be scanned by the a.) path option because of b.)

Please correct me if I have missed something in your issue

**Comment by w1ndy** [comment]
Thanks for the clarification! However I am still wondering if the
performance of reindexing can be benefited from scanning mounts only once,
for example, stop scanning /user2/Data because the same mount /user1/Data
has already been scanned. I guess this is achievable by mapping mount paths
back to some sort of unique ID for each external storage?

Also I'm not quite understand how to solve my case with the path option. Is
it sufficient to scan mounts for only one user, like /user1/Data, and other
users can reuse the cache generated?
Martin <notifications@github.com>于2018年8月1日 周三下午8:32写道：

**Comment by mmattel** [comment]
That is exactly what happens for sharing and federated shares, they are only scanned at the point of origin. The point of origin is the user which mounts the path and shares it to others. Only he needs to scan it and all others do participate on the result.

**Comment by w1ndy** [comment]
Thanks! Actually in my use case it seems that the point of origin does not
exist because the mount is a local external storage mapped from /some/path
on the host system. User admin can add the local mount in the external
storage panel and restrict the mount to a group of users, but I guess it is
not necessary to add user admin to that group so admin can view the mount,
too. So I'm wondering if scanning any user is sufficient in this scenario
to build the cache for the local external storage? Also hypothetically, is
it possible to skip external storage if I only want to rebuild cache for
all users' own files?

Martin <notifications@github.com>于2018年8月1日 周三下午9:11写道：

**Comment by jelinj8** [comment]
Hello,
I have another variant of this problem - I wanted almost all my users to have one folder with several mounted external storages.

1. I have used "group folders" for that - I've created a "mounts" group folder for my groups of users (those that see at least one mount)
1. mounted the external storages (local filesystem) into it (for sub-groups of users, e.g. I want my photos to be visible just inside my family, but audiobooks, movies and music are shared more widely)

It seems that the external storage is rescanned with the user's scan (a bad idea, as the external storage has about 4TB of media files and tooks ages).
